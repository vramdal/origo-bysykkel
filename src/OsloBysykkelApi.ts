import { AutoDiscoveryFeedType } from './Types';

export const URL_AUTODISCOVERY = 'https://gbfs.urbansharing.com/oslobysykkel.no/gbfs.json';

const CLIENT_IDENTIFIER = 'vidarramdal-origo-oppgave';

const apiFetch: (url: string) => Promise<Response> = async (url: string) => {
  return fetch(url.replace('http://', 'https://'), {
    headers: new Headers({
      'Client-Identifier': CLIENT_IDENTIFIER,
    }),
  });
};

export const fetchJson: <T>(url: string) => Promise<T> = async (url: string) => {
  return apiFetch(url).then((response) => response.json());
};

export function getFeedUrl(autoDiscoveryFeed: AutoDiscoveryFeedType, feedName: string): string {
  const match = autoDiscoveryFeed.data.nb.feeds.find((feed) => feed.name === feedName);
  if (!match) {
    throw new Error('Har ikke URL til feed: ' + feedName);
  }
  return match.url;
}

export const _testing = { apiFetch };
