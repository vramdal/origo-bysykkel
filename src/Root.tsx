import React from 'react';
import { Provider } from 'react-redux';
import App from './App';
import thunk from 'redux-thunk';
import { applyMiddleware, createStore } from 'redux';
import { reducer } from './Reducer';
import { composeWithDevTools } from 'redux-devtools-extension';

const middlewares = [thunk];

export const initializeStore = (): ReturnType<typeof createStore> => {
  return createStore(reducer, composeWithDevTools(applyMiddleware(...middlewares)));
};

export const RootComponent = (): JSX.Element => (
  <React.StrictMode>
    <Provider store={initializeStore()}>
      <App />
    </Provider>
  </React.StrictMode>
);
