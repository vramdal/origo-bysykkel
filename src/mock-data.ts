/* eslint-disable */
export const stationStatus1 = {
  last_updated: 1589788935,
  ttl: 10,
  data: {
    stations: [
      {
        station_id: '1755',
        is_installed: 1,
        is_renting: 1,
        is_returning: 1,
        last_reported: 1589788935,
        num_bikes_available: 11,
        num_docks_available: 22,
      },
      {
        station_id: '1101',
        is_installed: 1,
        is_renting: 1,
        is_returning: 1,
        last_reported: 1589788931,
        num_bikes_available: 5,
        num_docks_available: 19,
      },
    ],
  },
};

export const stationStatus2 = {
  last_updated: 1589788955,
  ttl: 10,
  data: {
    stations: [
      {
        station_id: '1755',
        is_installed: 1,
        is_renting: 1,
        is_returning: 1,
        last_reported: 1589788935,
        num_bikes_available: 0,
        num_docks_available: 33,
      },
      {
        station_id: '1101',
        is_installed: 1,
        is_renting: 1,
        is_returning: 1,
        last_reported: 1589788931,
        num_bikes_available: 24,
        num_docks_available: 0,
      },
    ],
  },
};

const URL_STATION_INFORMATION = 'http://gbfs.urbansharing.com/oslobysykkel.no/station_information.json';
const URL_STATION_STATUS = 'http://gbfs.urbansharing.com/oslobysykkel.no/station_status.json';

export const autoDiscoveryResponse = {
  last_updated: 1589749490,
  ttl: 10,
  data: {
    nb: {
      feeds: [
        {
          name: 'system_information',
          url: 'http://gbfs.urbansharing.com/oslobysykkel.no/system_information.json',
        },
        {
          name: 'station_information',
          url: URL_STATION_INFORMATION,
        },
        {
          name: 'station_status',
          url: URL_STATION_STATUS,
        },
      ],
    },
  },
};

export const stationInformationResponse = {
  last_updated: 1589788529,
  ttl: 10,
  data: {
    stations: [
      {
        station_id: '1755',
        name: 'Aker Brygge',
        address: 'Aker Brygge',
        lat: 59.91118372188379,
        lon: 10.730034556850455,
        capacity: 33,
      },
      {
        station_id: '1101',
        name: 'Stortingstunellen',
        address: 'R\u00e5dhusgata 34',
        lat: 59.91065301806209,
        lon: 10.737365277561025,
        capacity: 24,
      },
    ],
  },
};
