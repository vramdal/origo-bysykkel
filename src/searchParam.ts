export function useSearchParam<T>(
  searchParamName: string,
  location: Pick<typeof window.location, 'href' | 'search'>,
  defaultValue: string,
): [string, (newValue: string) => void] {
  const urlSearchParams = new URLSearchParams(location.search);
  const value = urlSearchParams.get(searchParamName) || defaultValue;
  const setValue = (newValue: string): void => {
    const url = new URL(location.href);
    url.searchParams.set(searchParamName, newValue);
    location.href = url.href;
  };
  return [value, setValue];
}
