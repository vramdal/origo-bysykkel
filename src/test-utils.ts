export function response(json: { last_updated: number; data: object; ttl: number }): Promise<object> {
  return Promise.resolve(json);
}

export const childNodesToText = (parent: HTMLElement | undefined | null): string | undefined =>
  (parent &&
    Array.from(parent.childNodes)
      .map((child) => child.textContent)
      .join(' | ')) ||
  undefined;
