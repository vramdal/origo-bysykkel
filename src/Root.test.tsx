import React from 'react';
import { render, wait, waitForElementToBeRemoved } from '@testing-library/react';
import { fetchJson } from './OsloBysykkelApi';
import { resetAllWhenMocks, when } from 'jest-when';
import { RootComponent } from './Root';
import { act } from 'react-dom/test-utils';
import { autoDiscoveryResponse, stationInformationResponse, stationStatus1, stationStatus2 } from './mock-data';
import { useSearchParam } from './searchParam';

jest.mock('./OsloBysykkelApi', () => {
  const originalModule = jest.requireActual('./OsloBysykkelApi');
  return {
    ...originalModule,
    fetchJson: jest.fn(),
  };
});

jest.mock('./searchParam');

const fetchJsonMock = fetchJson as jest.Mock;
const useSearchParamMock = useSearchParam as jest.Mock;

function response(json: { last_updated: number; data: object; ttl: number }): Promise<object> {
  return Promise.resolve(json);
}

const childNodesToText = (tr: HTMLElement): string =>
  Array.from(tr.children)
    .map((tr) => tr.textContent)
    .join(' | ');

describe('ui tests', () => {
  afterEach(() => {
    resetAllWhenMocks();
    jest.clearAllMocks();
  });

  beforeEach(() => {
    jest.useFakeTimers();
    when(fetchJsonMock)
      .calledWith(expect.stringContaining('gbfs.json'))
      .mockResolvedValue(response(autoDiscoveryResponse))
      .calledWith(expect.stringContaining('station_information.json'))
      .mockResolvedValue(response(stationInformationResponse));
  });

  describe('List rendering', () => {
    beforeEach(() => {
      when(useSearchParamMock)
        .calledWith('tab', expect.anything(), expect.anything())
        .mockReturnValue(['list', jest.fn()]);
    });

    test('it should render a list of stations', async () => {
      when(fetchJsonMock)
        .calledWith(expect.stringContaining('station_status.json'))
        .mockResolvedValueOnce(response(stationStatus1));
      const { getByTestId, container, getAllByRole } = render(<RootComponent />);

      expect(getByTestId('loading-indicator')).toBeInTheDocument();
      await waitForElementToBeRemoved(() => getByTestId('loading-indicator'));

      const rows = getAllByRole('row');
      expect(childNodesToText(rows[1])).toMatchInlineSnapshot(`"Aker Brygge | 22 | 11"`);
      expect(childNodesToText(rows[2])).toMatchInlineSnapshot(`"Stortingstunellen | 19 | 5"`);
      expect(container).toMatchSnapshot();
    });

    test('it should update the list with new numbers', async () => {
      when(fetchJsonMock)
        .calledWith(expect.stringContaining('station_status.json'))
        .mockResolvedValueOnce(response(stationStatus1))
        .calledWith(expect.stringContaining('station_status.json'))
        .mockResolvedValue(response(stationStatus2));

      const { getAllByRole, queryByTestId } = render(<RootComponent />);
      await wait(() => {
        expect(queryByTestId('loading-indicator')).toBeNull();
      });
      await act(async () => {
        await jest.runAllTimers();
      });

      await wait(() => {
        const rows = getAllByRole('row');
        expect(childNodesToText(rows[1])).toMatchInlineSnapshot(`"Aker Brygge | 33 | 0"`);
        expect(childNodesToText(rows[2])).toMatchInlineSnapshot(`"Stortingstunellen | 0 | 24"`);
      });
    });

    test('it should display an error message when station status fetch fails, then try again', async () => {
      when(fetchJsonMock)
        .calledWith(expect.stringContaining('station_status.json'))
        .mockRejectedValueOnce({ error: 'mock error' })
        .mockResolvedValue(response(stationStatus1));
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementationOnce(() => {});

      const { queryByTestId, getAllByRole } = render(<RootComponent />);
      await wait(() => {
        expect(queryByTestId('loading-indicator')).toBeNull();
      });

      await act(async () => {
        await jest.runAllTimers();
      });

      await wait(() => {
        const rows = getAllByRole('row');
        expect(childNodesToText(rows[1])).toMatchInlineSnapshot(`"Aker Brygge | 22 | 11"`);
        expect(childNodesToText(rows[2])).toMatchInlineSnapshot(`"Stortingstunellen | 19 | 5"`);
        expect(consoleErrorSpy).toHaveBeenCalledWith(expect.stringContaining('Error fetching stationStatus'), {
          error: 'mock error',
        });
      });
    });

    test('complete snapshot', async () => {
      when(fetchJsonMock)
        .calledWith(expect.stringContaining('station_status.json'))
        .mockResolvedValueOnce(response(stationStatus1));

      const { container, queryByTestId } = await render(<RootComponent />);
      await wait(() => {
        expect(queryByTestId('loading-indicator')).toBeNull();
      });

      expect(container).toMatchSnapshot();
    });
  });
});
