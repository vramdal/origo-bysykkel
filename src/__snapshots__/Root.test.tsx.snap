// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui tests List rendering complete snapshot 1`] = `
<div>
  <div
    class="state-indicator success-indicator"
    data-testid="success-indicator"
  />
  <div
    class="tab-bar"
  >
    <button
      role="button"
      type="button"
    >
      Map
    </button>
    <button
      disabled=""
      role="button"
      type="button"
    >
      List
    </button>
  </div>
  <div
    class="station-table"
    role="table"
  >
    <div
      class="row station-table__header"
      role="row"
    >
      <div
        class=""
        role="columnheader"
      >
        Stasjon
      </div>
      <div
        class=""
        role="columnheader"
      >
        Antall låser
      </div>
      <div
        class=""
        role="columnheader"
      >
        Antall sykler
      </div>
    </div>
    <div
      class="station-table__body"
      role="rowgroup"
    >
      <div
        class="row"
        role="row"
      >
        <div
          role="cell"
        >
          Aker Brygge
        </div>
        <div
          class=""
          role="cell"
        >
          22
        </div>
        <div
          class=""
          role="cell"
        >
          11
        </div>
      </div>
      <div
        class="row"
        role="row"
      >
        <div
          role="cell"
        >
          Stortingstunellen
        </div>
        <div
          class=""
          role="cell"
        >
          19
        </div>
        <div
          class=""
          role="cell"
        >
          5
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ui tests List rendering it should render a list of stations 3`] = `
<div>
  <div
    class="state-indicator success-indicator"
    data-testid="success-indicator"
  />
  <div
    class="tab-bar"
  >
    <button
      role="button"
      type="button"
    >
      Map
    </button>
    <button
      disabled=""
      role="button"
      type="button"
    >
      List
    </button>
  </div>
  <div
    class="station-table"
    role="table"
  >
    <div
      class="row station-table__header"
      role="row"
    >
      <div
        class=""
        role="columnheader"
      >
        Stasjon
      </div>
      <div
        class=""
        role="columnheader"
      >
        Antall låser
      </div>
      <div
        class=""
        role="columnheader"
      >
        Antall sykler
      </div>
    </div>
    <div
      class="station-table__body"
      role="rowgroup"
    >
      <div
        class="row"
        role="row"
      >
        <div
          role="cell"
        >
          Aker Brygge
        </div>
        <div
          class=""
          role="cell"
        >
          22
        </div>
        <div
          class=""
          role="cell"
        >
          11
        </div>
      </div>
      <div
        class="row"
        role="row"
      >
        <div
          role="cell"
        >
          Stortingstunellen
        </div>
        <div
          class=""
          role="cell"
        >
          19
        </div>
        <div
          class=""
          role="cell"
        >
          5
        </div>
      </div>
    </div>
  </div>
</div>
`;
